src/v/storage/offset_to_filepos_consumer.h

```diff
@@ -33,62 +33,39 @@ class offset_to_filepos_consumer {
     ss::future<ss::stop_iteration> operator()(::model::record_batch batch) {
         _prev = _accumulator;
         _accumulator += batch.size_bytes();
-        if (_target_last_offset == batch.base_offset()) {
-            _filepos = {batch.base_offset() - model::offset(1), _prev};
-            return ss::make_ready_future<ss::stop_iteration>(
-              ss::stop_iteration::yes);
-        }
-        // if we are in the middle of an actual batch, throw an exception to
-        // preserve previous semantics
 
+        if (_target_last_offset <= batch.base_offset()) {
+            // it may be the case that we hit this condition with first read
+            // batch in this case simply move to the offset preceding current
+            // batch base offset i.e. previous batch last offset
+            if (_prev_batch_last_offset == model::offset{}) {
+                _prev_batch_last_offset = model::prev_offset(
```

Comment:
I wonder if this is correct - if we started right inside the gap, presumably we want the last offset before the gap, not the "first offset after the gap" - 1? E.g. with batches [0,10] and [15,20] and if we start with the [15,20] batch and then truncate at 12, we will end with the segment ending at 10, but the offsets tracker will say that the last segment offset is 14.

Reply:
we are safe in this situation. We determine start offset based on the segment index. Index return first offset that is indexed and is smaller or equal to requested one. If we try to truncate in the gap we will always receive either exactly a batch with base offset being truncation point or the one from before a gap.

Reply:
As we agreed offline and @ztlpn pointed out there may be a case when a patch pointed by the segment index is a first batch after a gap